### PR TITLE
[DRAFT] - Subscriber performance 

### PR DESCRIPTION
**Pull request by kevburnsjr**
This PR is a draft. It's just here for discussion.   
This PR will be deleted and a new PR will be created against upstream once it is proven to fix the problem.

### CODE REVIEW

**Comment by divine** [comment]
Hello,

Well, first of all, thank you so much for taking a look and having this fixed.

This indeed fixes the problem I was having, it's now performing **amazingly fast** and without using much CPU, it's well under %8-%10.

However, there are some issues that I've observed.

If you don't have time, I'm ready to fix these issues and push PR towards this repository/branch - just let me know.

- Local transport doesn't deliver messages due to missing (`s.Ready()`):
https://github.com/kevburnsjr/mercure/blob/b302b59932abc33a27abfe8b8f12c2e5d6b1ff74/bolt_transport.go#L177-L196
in function https://github.com/kevburnsjr/mercure/blob/b302b59932abc33a27abfe8b8f12c2e5d6b1ff74/local_transport.go#L52-L68

- Caddy configuration still uses the old topic selector and it's not configurable:
https://github.com/kevburnsjr/mercure/blob/b302b59932abc33a27abfe8b8f12c2e5d6b1ff74/caddy/caddy.go#L139-L142

No other issues that I have found and again thank you so much!

Thanks!

**Comment by kevburnsjr** [comment]
Done and done. Tests are fixed to.

> amazingly fast and without using much CPU, it's well under %8-%10.

Got any graphs? 🚀

**Comment by divine** [comment]
> Done and done. Tests are fixed to.
> 
> > amazingly fast and without using much CPU, it's well under %8-%10.
> 
> Got any graphs? 🚀

Caddy file doesn't parse it, looks like it needs to parse algorithm after this case https://github.com/kevburnsjr/mercure/blob/subscriber-performance/caddy/caddy.go#L352:

```
case "lru_shard_size":
  if !d.NextArg() {
    return d.ArgErr()
  }
  
  v, err := strconv.ParseInt(d.Val(), 10, 64)
  if err != nil {
    return err //nolint:wrapcheck
  }
  
  m.LruShardSize = &v
```


Regarding graphs, do you mean `pprof` graphs, right? If yes, here we go:

kevfix


Again, thank you so much!

**Comment by divine** [comment]
Another graph after running for about 15 minutes:

```
# HELP mercure_subscribers The current number of running subscribers
# TYPE mercure_subscribers gauge
mercure_subscribers 2773
# HELP mercure_subscribers_total Total number of handled subscribers
# TYPE mercure_subscribers_total counter
mercure_subscribers_total 9155
# HELP mercure_updates_total Total number of handled updates
# TYPE mercure_updates_total counter
mercure_updates_total 64149

```

kevfix2

**Comment by kevburnsjr** [comment]
@divine I actually meant newrelic graphs or grafana or whatever shows the CPU usage dropping off a cliff like an action film.

I've never used or configured Caddy.  
I see [sample configs](https://github.com/kevburnsjr/mercure/blob/subscriber-performance/caddy/caddy_test.go#L171-L175) in the tests.   
I'm thinking a new directive:

```
			mercure {
				anonymous
				subscriptions
				publisher_jwt !ChangeMe!
				lru_cache 10000
			}
```

Makes sense to me. We can always change it when we switch to make lru the default and probably best to make it separate.

**Comment by divine** [comment]
@kevburnsjr sure, here we go:
Daily:
image
Weekly:
image

image

Yay, it's a really small virtual server that handles a few million notifications per day 😄.

> I'm thinking a new directive:

`lru_cache` makes sense! Maybe even shorter just `lru` but then it won't make sense probably.


Thanks!

**Comment by kevburnsjr** [comment]
@divine How do those numbers look? I was concerned they may have been creeping back up. Still looking good?

**Comment by divine** [comment]
Hello @kevburnsjr,

Well, after running for a day it was just stuck for no reason. I have double-checked the memory and CPU.

It was used both fully and notifications weren't dispatched to subscribed clients.

Probably the reason was a small server (EPYC 2 vCPU, 2GB RAM), I have spun up a new one (EPYC 3 vCPU, 4GB RAM) which looks a little better but still, CPU usage is a little bit high after running for a few hours.

I will report to you with my findings:

```
# HELP mercure_subscribers The current number of running subscribers
# TYPE mercure_subscribers gauge
mercure_subscribers 5154
# HELP mercure_subscribers_total Total number of handled subscribers
# TYPE mercure_subscribers_total counter
mercure_subscribers_total 726251
# HELP mercure_updates_total Total number of handled updates
# TYPE mercure_updates_total counter
mercure_updates_total 5.94848e+06
```

image

HTOP:
image

Caddy configuration:
`lru_shard_size 500000`

pprof:

kevfix3


[pprof.caddy.samples.cpu.030.pb.gz](https://github.com/kevburnsjr/mercure/files/7162908/pprof.caddy.samples.cpu.030.pb.gz)


Thanks!

**Comment by divine** [comment]
Increased the value of `lru_shard_size` to `1000000`

The result after the restart was fine then went up again due to high number of updates pushed to it:
image

Then went to normal back:


Almost idle:
image


image

pprof:
kevfix4
[pprof.caddy.samples.cpu.031.pb.gz](https://github.com/kevburnsjr/mercure/files/7163043/pprof.caddy.samples.cpu.031.pb.gz)


```bash
# HELP mercure_subscribers The current number of running subscribers
# TYPE mercure_subscribers gauge
mercure_subscribers 5282
# HELP mercure_subscribers_total Total number of handled subscribers
# TYPE mercure_subscribers_total counter
mercure_subscribers_total 17811
# HELP mercure_updates_total Total number of handled updates
# TYPE mercure_updates_total counter
mercure_updates_total 33347
```

Thanks!

**Comment by kevburnsjr** [comment]
It's pretty cool that you can serve 5k subscribers on a 3 core VM. I thought maybe to start looking for a memory leak but the fact that you've got CPU going up and down suggests it's probably not a leak.

The only leak I could find just by reading the code is here
https://github.com/dunglas/mercure/blob/main/subscribe.go#L107-L117

Where a subscriber is started, but the function returns nil so `Disconnect` is never called leaving the zombie subscription to sit and spin.
This error will always be seen in the logs as "Unable to add subscriber"
Do you see a lot of those?

**Comment by divine** [comment]
Hello!

> CPU going up and down suggests it's probably not a leak.

Exactly, it goes up and down when the high number of updates is pushed to it.

> Do you see a lot of those?

Actually, no such errors in the logs...

Thanks!

**Comment by kevburnsjr** [comment]
@divine I like mail.tm and I'd like to see it remain free. 
For that reason, I took another stab at improving efficiency. 

While the last set of enhancements was limited to improving the efficiency of the subscriber, this bit aims to improve the efficiency of message fan-out in the local transport. By caching a subscriber bitmap for each topic, we can reduce the number of cache lookups per update from `O(n)` to `O(1)` where `n` represents the number of subscribers. This new approach appears to be significantly more efficient for all workloads and can provide more than 20,000 times higher throughput for highly selective workloads with high concurrency. Care to give it a shot?

```
Before

cpu: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_1-2               6020            197048 ns/op
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_10-2              6756            178345 ns/op
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_100-2             6175            185047 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_1-2               679           1774345 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_10-2              680           1753074 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_100-2             664           1801593 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_1-2               134           8964758 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_10-2              132           9114793 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_100-2             129           9199351 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_1-2               58          19172861 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_10-2              56          19190180 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_100-2             61          18511490 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_1-2                331           3604988 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_10-2               337           3533874 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_100-2              606           1966361 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_1-2                32          35837120 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_10-2               33          35347438 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_100-2              57          19961933 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_1-2                 6         181751742 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_10-2                6         178616568 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_100-2              12          98426881 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_1-2                3         362162971 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_10-2               3         362948712 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_100-2              6         198618388 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_1-2                82          14360454 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_10-2               85          13955853 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_100-2             141           8413187 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_1-2                8         143085475 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_10-2               8         140071410 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_100-2             13          84186038 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_1-2                2         723826206 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_10-2               2         693173880 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_100-2              3         418398172 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_1-2               1        1405078617 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_10-2              1        1375515850 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_100-2             2         846991015 ns/op
PASS
```
```
After

cpu: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_1-2             264186              3947 ns/op
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_10-2            155938              7828 ns/op
BenchmarkLocalTransport/topics_1_concurrency_100_matchpct_100-2            25880             46677 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_1-2            148768              8461 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_10-2            25450             47901 ns/op
BenchmarkLocalTransport/topics_1_concurrency_1000_matchpct_100-2            3087            469505 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_1-2             40352             29910 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_10-2             8838            238562 ns/op
BenchmarkLocalTransport/topics_1_concurrency_5000_matchpct_100-2             546           2338087 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_1-2            21139             57396 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_10-2            3264            488230 ns/op
BenchmarkLocalTransport/topics_1_concurrency_10000_matchpct_100-2            262           5200914 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_1-2             195138              6664 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_10-2            118936             10112 ns/op
BenchmarkLocalTransport/topics_5_concurrency_100_matchpct_100-2            25681             47226 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_1-2            117304             10803 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_10-2            24054             48372 ns/op
BenchmarkLocalTransport/topics_5_concurrency_1000_matchpct_100-2            3770            469121 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_1-2             50228             30129 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_10-2             5155            235773 ns/op
BenchmarkLocalTransport/topics_5_concurrency_5000_matchpct_100-2             721           2422102 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_1-2            19496             61929 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_10-2            2468            481776 ns/op
BenchmarkLocalTransport/topics_5_concurrency_10000_matchpct_100-2            274           5617101 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_1-2            118710             10484 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_10-2            84920             13410 ns/op
BenchmarkLocalTransport/topics_10_concurrency_100_matchpct_100-2           24715             49139 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_1-2            87052             14284 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_10-2           23744             51394 ns/op
BenchmarkLocalTransport/topics_10_concurrency_1000_matchpct_100-2           2560            470790 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_1-2            36888             32576 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_10-2            5348            243063 ns/op
BenchmarkLocalTransport/topics_10_concurrency_5000_matchpct_100-2            505           2365070 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_1-2           19935             60491 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_10-2           2673            487560 ns/op
BenchmarkLocalTransport/topics_10_concurrency_10000_matchpct_100-2           200           5642333 ns/op
PASS
ok      github.com/dunglas/mercure      125.328s
```

**Comment by divine** [comment]
Hello @kevburnsjr,

Again, I'm really grateful for the amazing work you've done again and I'm happy that you liked our service.

I've gone ahead and just deployed your new fix but it's leaking after some time.

Before deployment:
image

After:

image


Heap:
heap

Allocs:
allocs

CPU:
cpu

Files:
[pprof.caddy.heap.pb.gz](https://github.com/kevburnsjr/mercure/files/7234569/pprof.caddy.heap.pb.gz)
[pprof.caddy.samples.cpu.pb.gz](https://github.com/kevburnsjr/mercure/files/7234570/pprof.caddy.samples.cpu.pb.gz)
[pprof.caddy.allocs.pb.gz](https://github.com/kevburnsjr/mercure/files/7234571/pprof.caddy.allocs.pb.gz)

This is with "lru_shard_size 100000" configuration, I've just commented out and the results are the same:

image

I will try to redeploy it again, maybe I've made some mistakes (happens sometimes) but this is the result what I've got.

Thanks!

**Comment by kevburnsjr** [comment]
@divine Okay, looks like a subscriber leak which makes sense. 
`subscriber.Disconnect()` was sufficient to remove references to the disconnected subscriber before but now 
`transport.RemoveSubscriber(s)` must be called explicitly during shutdown. Try it now?

**Comment by kevburnsjr** [comment]
Ooops, fixing one sec

**Comment by kevburnsjr** [comment]
There we go. Thought I'd already added the new `RemoveSubscriber` method to the Transport interface. Should build now.

**Comment by divine** [comment]
@kevburnsjr I don't know how to explain this but the result looks amazing...

image

Just 10 minutes result:
```bash
# HELP mercure_subscribers The current number of running subscribers
# TYPE mercure_subscribers gauge
mercure_subscribers 5131
# HELP mercure_subscribers_total Total number of handled subscribers
# TYPE mercure_subscribers_total counter
mercure_subscribers_total 8835
# HELP mercure_updates_total Total number of handled updates
# TYPE mercure_updates_total counter
mercure_updates_total 40847
```

And finally, the performance issue has been fixed completely, thank you so much!

**Comment by kevburnsjr** [comment]
@divine Nice! 

I'll check back in a couple of days and if it remains stable then I'll parameterize the skip filter size (currently hard coded at `1e5`) and submit a PR to the upstream repo next weekend. 👍 

Let me know if you run into any issues.
If memory usage is too high you can tune [this hard coded param](https://github.com/kevburnsjr/mercure/blob/2224cb7c95d4573167598bfa40ae52898b82f9ea/local_transport.go#:~:text=1e5) down to `1e4`.

**Comment by dunglas** [comment]
Wow, thank you very much for your work on this @kevburnsjr! The results look impressive!

File: caddy/caddy.go [hunk 1]
```diff
@@ -136,9 +139,18 @@ func (m *Mercure) Provision(ctx caddy.Context) error { //nolint:funlen
...
-		return err //nolint:wrapcheck
+	var err error
+	var tss *mercure.TopicSelectorStore
+	if m.LruShardSize != nil {
+		tss, err = mercure.NewTopicSelectorStoreLru(*m.LruShardSize, mercure.DefaultTopicSelectorStoreLruShardCount)
+		if err != nil {
+			return err //nolint:wrapcheck
+		}
+	} else {
+		tss, err = mercure.NewTopicSelectorStore(nc, mc)
```
Comment (dunglas) [inline caddy/caddy.go:150]: Is there any benefit to keeping the old implementation?

File: caddy/caddy.go [hunk 2]
```diff
@@ -90,6 +90,9 @@ type Mercure struct {
 	// Maximum cache cost, defaults to 100MB, set to -1 to disable the cache. See https://github.com/dgraph-io/ristretto for details.
 	CacheMaxCost *int64 `json:"cache_max_cost,omitempty"`
 
+	// Triggers use of LRU topic selector cache and avoidance of select priority queue (recommend 10,000 - 1,000,000)
+	LruShardSize *int64 `json:"lru_shard_size,omitempty"`
```
Comment (dunglas) [inline caddy/caddy.go]:
```suggestion
	LRUShardSize *int64 `json:"lru_shard_size,omitempty"`
```

IIRC our CS.

**Comment by kevburnsjr** [comment]
Some additional context to validate the approach...

This whitepaper from ZeroMQ (2007) describes a very similar approach in their use of inverted bitmaps to handle wildcard topic routing.
http://wiki.zeromq.org/whitepapers:message-matching

This article from AMQP is basically the same article with some more info about their implementation.
http://openamq.wikidot.com/doc:fast-topic-matching#toc8

> The maximum number of criteria is around a hundred thousand.

We even arrived at the same defaults 😄 

No magic here, just some good old fashioned computer science 👍

File: subscriber.go [hunk 3]
```diff
@@ -168,46 +98,78 @@ func (s *Subscriber) Receive() <-chan *Update {
 // HistoryDispatched must be called when all messages coming from the history have been dispatched.
 func (s *Subscriber) HistoryDispatched(responseLastEventID string) {
```
Comment (dunglas) [inline subscriber.go:99]: Shouldn't we remove this method in favor of `Ready()`?

**Comment by dunglas** [comment]
Hi @kevburnsjr!

I finally had the time to review this PR in depth. Excellent work!!  Thank you.
I proposed some tiny changes and fixes in https://github.com/kevburnsjr/mercure/pull/2, but overall I think that we can merge this one!

Can you open the PR?

Best regards.

**Comment by kevburnsjr** [comment]
@dunglas Sounds good. I'll open a PR against upstream this weekend.

**Comment by divine** [comment]
Hello,

I haven't replied since everything running quite stable with zero issues.

The only question I have, shouldn't be there only one topic store instead of two but this is probably up to @dunglas to decide...

Thanks!

**Comment by dunglas** [comment]
I'm all for keeping only one store.

**Pull request closed without merging**